include/picongpu/plugins/xrayDiffraction/ComputeLocalDomain.kernel

```diff
@@ -0,0 +1,359 @@
+
+#include <pmacc/dimensions/DataSpaceOperations.hpp>
+#include <pmacc/mappings/threads/ForEachIdx.hpp>
+#include <pmacc/mappings/threads/IdxConfig.hpp>
+#include <pmacc/math/Complex.hpp>
+#include <pmacc/memory/Array.hpp>
+#include <pmacc/memory/shared/Allocate.hpp>
+
+#include <cstdint>
+
+
+namespace picongpu
+{
+namespace plugins
+{
+namespace xrayDiffraction
+{
+namespace detail
+{
+
+    //! Complex type to represent structure factor values
+    using Complex = pmacc::math::Complex< float_X >;
+
+    //! Internal representation of results for each thread
+    struct Result
+    {
+        //! Structure factor value F(q) for a single scattering vector
+        Complex structureFactor = Complex::zero();
+        
+        //! Number of processed macroparticles
+        uint64_t numMacroparticles = 0u;
+
+        //! Combined weighting of processed macroparticles
+        float_64 totalWeighting = 0.0;
+        
+        //! Combine the current result with another one
+        HDINLINE Result & operator+=( Result const & other )
+        {
+            structureFactor += other.structureFactor;
+            numMacroparticles += other.numMacroparticles;
+            totalWeighting += other.totalWeighting;
+            return *this;
+        }
+
+    };
+ 
+    /** Process particles of a frame
+     *
+     * @tparam T_numWorkers number of workers
+     * @tparam T_Acc alpaka accelerator type
+     * @tparam T_Frame particle frame type
+     * @tparam T_CachedGlobalPositionArray cached array of global positions type
+     * @tparam T_CachedWeightingArray cached array of weighting type
+     *
+     * @param acc alpaka accelerator
+     * @param frame particle frame
+     * @param numParticles number of macroparticles in the frame
+     * @param particleBox particle box
+     * @param cachedGlobalPosition global positions array in shared memory
+     * @param cachedWeighting weighting array in shared memory
+     * @param supercellOffsetGlobal supercell offset in the global domain
+     * @param reciprocalSpace reciprocal space of scattering vectors
+     */
+    template<
+        uint32_t T_numWorkers,
+        typename T_Acc,
+        typename T_Frame,
+        typename T_CachedGlobalPositionArray,
+        typename T_CachedWeightingArray
+    >
+    DINLINE Result processFrame(
+        T_Acc const & acc,
+        T_Frame const & frame,
+        uint32_t const numParticles,
+        T_CachedGlobalPositionArray & cachedGlobalPosition,
+        T_CachedWeightingArray & cachedWeighting,
+        DataSpace< simDim > const & supercellOffsetGlobal,
+        ReciprocalSpace const & reciprocalSpace
+    )
+    {
+        using namespace pmacc::mappings::threads;
+        uint32_t const workerIdx = threadIdx.x;
+        uint32_t const globalWorkerIdx = blockIdx.x * blockDim.x + threadIdx.x;
+
+        // Collectively load frame data to the cached arrays
+        /// TODO: fix computation to get size from particle box/frame, not supercell size
```

Comment:
@psychocoderHPC we discussed it offline, but I forgot how to do it: I would like to get the frame size (for full frames) from the particles box/frame type, but not from the supercell, to not rely on them being the same as right now. How to accomplish it?

Reply:
`constexpr uint32_t frameSize = math::CT::volume< typename T_Frame::SuperCellSize >::type::value;`

Reply:
Thanks, I actually found it but this again says `SuperCellSize`, so I thought I needed something else. I think another variable should be added (in another PR of course) that more explicitly says that it's frame capacity.

Reply:
Fixed now.